gpbft/participant.go

```diff
@@ -202,36 +201,52 @@ func (p *Participant) beginInstance() error {
 	return nil
 }
 
-func (p *Participant) getCommittee(instance uint64) (*committee, error) {
-	c, ok := p.committees[instance]
+// Fetches the committee against which to validate messages for some instance.
+func (p *Participant) fetchCommittee(instance uint64) (*committee, error) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	// Reject messages for past instances.
+	if instance < p.currentInstance {
+		return nil, fmt.Errorf("instance %d, current %d: %w",
+			instance, p.currentInstance, ErrValidationTooOld)
+	}
+
+	comt, ok := p.committees[instance]
 	if !ok {
 		power, beacon, err := p.host.GetCommitteeForInstance(instance)
 		if err != nil {
-			return nil, fmt.Errorf("failed fetching committee for instance %d: %w", instance, err)
+			return nil, fmt.Errorf("instance %d: %w: %w", instance, ErrValidationNoCommittee, err)
 		}
 		if err := power.Validate(); err != nil {
-			return nil, fmt.Errorf("invalid power table: %w", err)
+			return nil, fmt.Errorf("instance %d: %w: invalid power: %w", instance, ErrValidationNoCommittee, err)
 		}
-		c = &committee{power: power, beacon: beacon}
-		p.committees[instance] = c
+		comt = &committee{power: power, beacon: beacon}
+		p.committees[instance] = comt
 	}
-	return c, nil
+	return comt, nil
 }
 
 func (p *Participant) handleDecision() {
 	if p.terminated() {
-		p.finalised = p.gpbft.terminationValue
-		p.terminatedDuringRound = p.gpbft.round
-		delete(p.committees, p.gpbft.instanceID)
-		p.gpbft = nil
-		p.currentInstance++
+		p.finishCurrentInstance()
 		nextStart := p.host.ReceiveDecision(p.finalised)
-
 		// Set an alarm at which to fetch the next chain and begin a new instance.
 		p.host.SetAlarm(nextStart)
 	}
 }
 
+func (p *Participant) finishCurrentInstance() {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	p.finalised = p.gpbft.terminationValue
+	p.terminatedDuringRound = p.gpbft.round
+	delete(p.committees, p.gpbft.instanceID)
+	p.gpbft = nil
+	p.currentInstance++
```

Comment:
The only things we actually _need_ to lock are `p.committees` and `p.currentInstance`, right? I'd consider putting everything else outside of the lock for clarity.